builder/replica_pipeline.cc

```diff
@@ -0,0 +1,640 @@
+
+  template <typename T, typename std::enable_if<std::is_integral<T>::value>::type* = nullptr>
+  void operator()(T value) {
+    result += std::to_string(value);
+  }
+
+  std::string result;
+};
+
+template <typename... Args>
+std::string str_cat(Args&&... args) {
+  size_t total_size = (ConcatHelper::size(std::forward<Args>(args)) + ...);
+
+  ConcatHelper concat_helper;
+  concat_helper.result.reserve(total_size);
+  (concat_helper(std::forward<Args>(args)), ...);
+
+  return concat_helper.result;
+}
+
+class pipeline_replicator : public recursive_node_visitor {
+public:
+  explicit pipeline_replicator(node_context& ctx) : ctx_(ctx) {}
+
+  void fail(const char* msg) {
+    std::cerr << "Unimplemented/TODO: " << msg << "\n";
+    std::abort();
+  }
+
+  void visit(const variable* op) override {
+    const auto& name = ctx_.name(op->sym);
+
+    if (buffers_emitted_.count(op->sym)) {
+      auto it = buffer_variables_emitted_.find(op->sym);
+      if (it != buffer_variables_emitted_.end()) {
+        name_ = it->second;
+      } else {
+        name_ = print_assignment_prefixed("_", "variable::make(" + name + "->sym())");
+        buffer_variables_emitted_[op->sym] = name_;
+      }
+      return;
+    }
+
+    name_ = print(var(op->sym));
+  }
+
+  void visit(const constant* op) override { name_ = std::to_string(op->value); }
+  void visit(const let* op) override { fail("unimplemented let"); }
+  void visit(const add* op) override { visit_binary_op(op, "+"); }
+  void visit(const sub* op) override { visit_binary_op(op, "-"); }
+  void visit(const mul* op) override { visit_binary_op(op, "*"); }
+  void visit(const div* op) override { visit_binary_op(op, "/"); }
+  void visit(const mod* op) override { visit_binary_op(op, "%"); }
+  void visit(const class min* op) override { visit_binary_call(op, "min"); }
+  void visit(const class max* op) override { visit_binary_call(op, "max"); }
+  void visit(const equal* op) override { visit_binary_op(op, "=="); }
+  void visit(const not_equal* op) override { visit_binary_op(op, "!="); }
+  void visit(const less* op) override { visit_binary_op(op, "<"); }
+  void visit(const less_equal* op) override { visit_binary_op(op, "<="); }
+  void visit(const logical_and* op) override { visit_binary_op(op, "&&"); }
+  void visit(const logical_or* op) override { visit_binary_op(op, "||"); }
+  void visit(const logical_not* op) override {
+    auto s = print_expr_maybe_inlined(op->a);
+    name_ = "!(" + s + ")";
+  }
+  void visit(const class select* op) override { fail("unimplemented select"); }
+
+  void visit(const call* op) override {
+    std::string call_name;
+    switch (op->intrinsic) {
+    case intrinsic::positive_infinity: call_name = "positive_infinity"; break;
+    case intrinsic::negative_infinity: call_name = "negative_infinity"; break;
+    case intrinsic::indeterminate: call_name = "indeterminate"; break;
+    case intrinsic::abs:
+      os_ << "  using std::abs;\n";
```

Comment:
We could do that but it would require a second visitor pass or other restructuring of the code to do this on-demand. (And I'm pretty sure we don't want to pre-emptively emit it just in case, since none of our current tests execute this.) I can make it so if you like, it will just add complexity to the code that I'm not sure pays for itself.

Reply:
I think just doing it unconditionally even if abs, min, max etc. are unused is fine. In that case, this is very simple (just a constant string that you initialize the output to instead of an empty string?). I think those are the only 3 things we need, if it really grows to the point that it is unwieldy we can do something else then.

Reply:
So to be clear, you want everything we emit to begin with

```
[]() -> ::slinky::pipeline {
  using std::abs, std::min , std::max;  <-- new
  node_context ctx;
}
```

Reply:
Yes, that's what I was imagining (I was picturing each `using` on one line, but what you have is fine too).

Reply:
(I only mention to illustrate that even if the preamble was bigger, I still think it's fine)